File: IntersectionOfTwoSortedArrays.c
#include <stdio.h>

// Hàm tìm giao của hai mảng đã được sắp xếp
void findIntersection(int a[], int b[], int m, int n) {
    int i = 0, j = 0;

    printf("Phần tử chung của hai mảng là: ");
    while (i < m && j < n) {
        // Bỏ qua phần tử trùng lặp trong mảng a
        if (i > 0 && a[i] == a[i - 1]) {
            i++;
            continue;
        }
        // So sánh và tìm phần tử chung
        if (a[i] < b[j]) {
            i++;
        } else if (a[i] > b[j]) {
            j++;
        } else {
            printf("%d ", a[i]);
            i++;
            j++;
        }
    }
    printf("\n");
}

int main() {
    int m, n;

    // Nhập mảng a
    printf("Nhập số phần tử của mảng a: ");
    scanf("%d", &m);
    int a[m];
    printf("Nhập %d phần tử của mảng a (đã sắp xếp): ", m);
    for (int i = 0; i < m; i++) {
        scanf("%d", &a[i]);
    }

    // Nhập mảng b
    printf("Nhập số phần tử của mảng b: ");
    scanf("%d", &n);
    int b[n];
    printf("Nhập %d phần tử của mảng b (đã sắp xếp): ", n);
    for (int i = 0; i < n; i++) {
        scanf("%d", &b[i]);
    }

    // Gọi hàm tìm giao của hai mảng
    findIntersection(a, b, m, n);

    return 0;
}

File: IntersectionOfTwoSortedArrays.py
def intersection(a, b):
    res = []  # Danh sách để lưu kết quả giao
    m = len(a)  # Độ dài của mảng a
    n = len(b)  # Độ dài của mảng b
    # Sử dụng thuật toán hai con trỏ để tìm giao
    i, j = 0, 0
    while i < m and j < n:
        # Bỏ qua các phần tử trùng lặp trong mảng a
        if i > 0 and a[i - 1] == a[i]:
            i += 1
            continue
        # So sánh phần tử tại a[i] và b[j]
        if a[i] < b[j]:  # Nếu a[i] nhỏ hơn b[j], tăng con trỏ i
            i += 1
        elif a[i] > b[j]:  # Nếu a[i] lớn hơn b[j], tăng con trỏ j
            j += 1
        else:  # Nếu a[i] bằng b[j], thêm vào danh sách kết quả
            res.append(a[i])
            i += 1
            j += 1
    return res  # Trả về danh sách giao
if __name__ == "__main__":
    
    a = list(map(int, input("Nhập phần tử của mảng a (đã sắp xếp): ").split()))
    b = list(map(int, input("Nhập phần tử của mảng b (đã sắp xếp): ").split()))
    res = intersection(a, b)
    print("Các phần tử chung của hai mảng là: ", " ".join(map(str, res)))
